spec/lib/alerts/de_userfying_edit_alert_monitor_spec.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require "#{Rails.root}/lib/alerts/de_userfying_edit_alert_monitor"
+
+def mock_mailer
+  OpenStruct.new(deliver_now: true)
+end
+
+describe DeUserfyingEditAlertMonitor do
+  let(:mntor) { described_class.new }
+  let(:course1) { create(:course, slug: 'slug-one') }
+  let(:course2) { create(:course, slug: 'slug-two') }
+  let(:student1) { create(:user, username: 'alice', email: 'student1@example.edu') }
+  let(:student2) { create(:user, username: 'bob', email: 'student2@example.edu') }
+  let(:student3) { create(:user, username: 'carol', email: 'student3@example.edu') }
+  let(:article1) { create(:article, title: 'my title 1', mw_page_id: 111) }
+  let(:article2) { create(:article, title: 'my title 2', mw_page_id: 222) }
+  let(:content_expert) { create(:user, greeter: true) }
+  let!(:student_role) { CoursesUsers::Roles::STUDENT_ROLE }
+  let(:courses_users) do
+    [
+      { course_id: course1.id, user_id: student1.id, role: student_role },
+      { course_id: course1.id, user_id: student3.id, role: student_role },
+      { course_id: course2.id, user_id: student2.id, role: student_role },
+      { course_id: course2.id, user_id: student3.id, role: student_role }
+    ]
+  end
+
+  before do
+    populate_students
+  end
+
+  describe '.edits' do
+    it 'checks keys' do
+      VCR.use_cassette 'recent_changes' do
+        fedits = mntor.edits.first
+        expect(fedits.dig('logparams', 'target_title')).not_to be_nil
+        expect(fedits.dig('user')).not_to be_nil
+        expect(fedits.dig('revid')).not_to be_nil
+        expect(fedits.dig('pageid')).not_to be_nil
+      end
+    end
+  end
+
+  describe '.current_students' do
+    it 'checks distinct enrolled students' do
+      expect(mntor.current_students.count).to eq 3
+    end
+  end
+
+  describe '.edits_made_by_students' do
+    it 'filters students that de-userfyed their page' do
+      allow(mntor).to receive(:edits).and_return(editsfeed)
+      edits = mntor.edits
+      students = mntor.current_students
+      expect(mntor.edits_made_by_students(edits, students).count).to eq editsfeed.count
+    end
+  end
+
+  describe '.create_alerts' do
+    before do
+      populate_content_expert
+      populate_articles_courses
```

Comment:
What happens if neither the Articles nor ArticlesCourses records exist when the monitor runs?

Reply:
Oops, I forgot to follow your advice not to use ArticlesCourses from the last conversation. (Actually, did look at it locally but then forgot :) )
I also add the case where article from the wiki not present in the dashboard DB.